src/syntaxtree.c

```diff
@@ -0,0 +1,1459 @@
+
+#include <ctype.h>
+
+typedef UInt4 LVar;
+typedef UInt4 HVar;
+typedef UInt GVar;
+
+typedef Obj (*CompileExpr)(Expr expr);
+
+typedef struct {
+  UInt tnum;
+  CompileExpr compiler;
+  const Char *name;
+  UInt arity;
+  const Char *argnames[6];
+} ExprT;
+
+static const ExprT AllExpressions[];
+
+static inline Obj NewSyntaxTreeNode(const char *type)
+{
+    Obj result;
+    Obj typestr;
+
+    C_NEW_STRING_DYN(typestr, type);
+    result = NEW_PREC(1);
+    AssPRec(result, RNamName("type"), typestr);
+
+    return result;
+}
+
+static Obj SyntaxTreeDefaultExpr(Expr expr)
+{
+    Obj result;
+
+    /* TODO: Check that TNUM in range */
+
+    result = NewSyntaxTreeNode(AllExpressions[TNUM_EXPR(expr)].name);
+    return result;
+}
+
+static inline Obj SyntaxTreeExpr(Expr expr)
+{
+    return AllExpressions[TNUM_EXPR(expr)].compiler(expr);
+}
+
+Obj SyntaxTreeUnknownExpr(Expr expr)
+{
+    Obj result;
+
+    result = NewSyntaxTreeNode("UnknownExpr");;
+    AssPRec(result, RNamName("tnum"), INTOBJ_INT(TNUM_EXPR(expr)));
+    return result;
+}
+
+Obj SyntaxTreeUnknownBool(Expr expr)
+{
+    /* compile the expression and check that the value is boolean          */
+    /* TODO: Check boolean? */
+    return SyntaxTreeExpr(expr);
+}
+
+extern Obj SyntaxTreeRefGVarFopy (Expr expr);
+
+Obj SyntaxTreeFunccall(Expr expr)
+{
+    Obj result;
+    Obj func;
+    Obj args, argi;
+    UInt narg, i;
+
+    result = NewSyntaxTreeNode("Funccall");;
+    /* TODO: If this is a gvar ref, put name? */
+    func = SyntaxTreeExpr( FUNC_CALL(expr) );
+
+    /*
+    if ( TNUM_EXPR( FUNC_CALL(expr) ) == T_REF_GVAR ) {
+    } */
+    AssPRec(result, RNamName("function"), func);
+
+    /* compile the argument expressions                                    */
+    narg = NARG_SIZE_CALL(SIZE_EXPR(expr));
+    args = NEW_PLIST(T_PLIST, narg);
+    SET_LEN_PLIST(args, narg);
+
+    for ( i = 1; i <= narg; i++ ) {
+        argi = SyntaxTreeExpr( ARGI_CALL( expr, i ) );
+        SET_ELM_PLIST(args, i, argi);
+        CHANGED_BAG(args);
+    }
+    AssPRec(result, RNamName("args"), args);
+    return result;
+}
+
+/* TODO: FuncCall options */
+Obj SyntaxTreeFunccallOpts(Expr expr)
+{
+    Obj result;
+    Obj opts;
+
+    result = NewSyntaxTreeNode("FunccallOpts");;
+    opts = SyntaxTreeExpr(ADDR_STAT(expr)[0]);
+    AssPRec(result, RNamName("opts"), opts);
+
+    return result;
+}
+
+static Obj SyntaxTreeFunc(Obj func);
+
+/* TODO: again? */
+Obj SyntaxTreeFuncExpr(Expr expr)
+{
+    Obj result;
+    Obj fexs;
+    Obj fexp;
+    Obj func;
+
+    result = NewSyntaxTreeNode("FuncExpr");;
+
+    fexs = FEXS_FUNC(CURR_FUNC);
+    fexp = ELM_PLIST(fexs, ((Int *)ADDR_EXPR(expr))[0]);
+    func = SyntaxTreeFunc(fexp);
+
+    AssPRec(result, RNamName("func"), func);
+
+    return result;
+}
+
+Obj SyntaxTreeIntExpr(Expr expr)
+{
+    Obj result;
+    Obj value;
+    Int siz;
+    UInt typ;
+    Int i;
+
+    result = NewSyntaxTreeNode("IntExpr");;
+
+    if(IS_INTEXPR(expr)) {
```

Comment:
Never noticed this before, but: `T_INT_EXPR` vs. `T_INTEXPR` seems needlessly confusing; perhaps we should rename one or both, to indicate one is immediate / small, the other big?